### PR TITLE
Feat/Introducing sidecars - Fluentbit & Firelens

### PR DESCRIPTION
**Pull request by rokibulislaam**

### CODE REVIEW

File: cloudlift/config/service_configuration.py [hunk 1]
```diff
@@ -229,9 +238,80 @@ def _validate_changes(self, configuration):
                 },
                 "logging": {
                     "oneOf": [
-                        {"type": "string", "pattern": "^(awslogs|fluentd|null)$"},
+                        {"type": "string", "pattern": "^(awslogs|awsfirelens|null)$"},
```
Comment (krishnanand275) [inline cloudlift/config/service_configuration.py:241]: Lets keep default awslogs untill we have full release.  This is avoid  issues during prod deployments. @shoan what you think ?

File: cloudlift/deployment/service_template_generator.py [hunk 2]
```diff
@@ -959,6 +981,71 @@ def _get_desired_task_count_for_service(self, service_name):
             return self.desired_counts[service_name]
         else:
             return 0
+    def _firelens_container_def_args_override(self, configuration, container_def_args):
+        if configuration.get('logging') == 'awsfirelens':
+            container_def_args['DependsOn'] = container_def_args.get('DependsOn', []) + [ContainerDependency(ContainerName='fluentbit-firelens-sidecar', Condition='START')]
```
Comment (krishnanand275) [inline cloudlift/deployment/service_template_generator.py:986]: Lets add a condition to check if sidecar container exists, also check if condition will be "START" or "healthy" @rokibulislaam

File: cloudlift/deployment/service_template_generator.py [hunk 3]
```diff
@@ -302,16 +319,18 @@ def _add_service(self, service_name, config):
                     RootDirectory=config['volume']['efs_directory_path']
                 )
             )]
+        if 'env' in config and isinstance(config.get('env'), dict):
```
Comment (krishnanand275) [inline cloudlift/deployment/service_template_generator.py:322]: Is this required ? @rokibulislaam

File: cloudlift/deployment/service_template_generator.py [hunk 4]
```diff
@@ -959,6 +981,71 @@ def _get_desired_task_count_for_service(self, service_name):
...
+            for index, sidecar in enumerate(configuration['sidecars']):
+                container_name = sidecar.get('name', f"sidecar_{index + 1}")
+                if not container_name.endswith('-sidecar'):
+                    container_name = container_name + '-sidecar'
+                image_uri = sidecar.get('image_uri')
+                memory_reservation = int(sidecar.get('memory_reservation', 50))
+                essential = sidecar.get('essential', True)
+                
+                mount_points = []
+                if sidecar.get('mount_points') and isinstance(sidecar['mount_points'], list):
```
Comment (krishnanand275) [inline cloudlift/deployment/service_template_generator.py:1001]: Let review if we need this  @rokibulislaam

File: cloudlift/deployment/service_template_generator.py [hunk 5]
```diff
@@ -959,6 +981,71 @@ def _get_desired_task_count_for_service(self, service_name):
...
+                    container_def_args['Command'] = [command]
+                    
+                env = sidecar.get('env')
+                if env is not None and isinstance(env, dict):
+                    container_def_args['Environment'] = []
+                    for key, value in env.items():
+                        container_def_args['Environment'].append(Environment(Name=key, Value=value))
+                logging = sidecar.get('logging')
+                if logging is not None:
+                    container_def_args['LogConfiguration'] = self._gen_log_config(service_name + f"-{container_name}", "awslogs" if 'logging' not in sidecar else logging)
```
Comment (krishnanand275) [inline cloudlift/deployment/service_template_generator.py:1020]: By default sidecar container , should have logging as awsfirelens, unless it is a fluentbit sidecar @rokibulislaam

**Review by krishnanand275 — CHANGES_REQUESTED** [review]

File: cloudlift/config/service_configuration.py [hunk 6]
```diff
@@ -257,6 +337,10 @@ def _validate_changes(self, configuration):
             "required": ["cloudlift_version", "services", "notifications_arn"]
         }
         try:
+            for _, service_configuration in configuration.get('services').items():
+                for sidecar in service_configuration.get('sidecars', []):
+                    if sidecar.get('name') == 'fluentbit-firelens-sidecar' and sidecar.get('log_driver') == 'awsfirelens':
```
[the remark this answers is not in the document]
Comment (rokibulislaam) [reply cloudlift/config/service_configuration.py:342]: Taken care of. Now it has a single source of truth from `cloudlift/constants.py`

File: cloudlift/deployment/service_template_generator.py [hunk 7]
```diff
@@ -302,16 +319,18 @@ def _add_service(self, service_name, config):
                     RootDirectory=config['volume']['efs_directory_path']
                 )
             )]
+        if 'env' in config and isinstance(config.get('env'), dict):
+            launch_type_td['Environment'] = [Environment(Name=k, Value=v) for (k, v) in config.get('env').items()]
```
Comment (rokibulislaam) [inline cloudlift/deployment/service_template_generator.py:323]: Not needed! YAGNI

File: cloudlift/deployment/service_template_generator.py (hunk 3 above)
Comment (rokibulislaam) [reply cloudlift/deployment/service_template_generator.py:322]: Not needed!

File: cloudlift/config/service_configuration.py (hunk 1 above)
Comment (shoan) [reply cloudlift/config/service_configuration.py:242]: We should keep awslogs as default.

File: cloudlift/config/service_configuration.py [hunk 8]
```diff
@@ -257,6 +338,10 @@ def _validate_changes(self, configuration):
             "required": ["cloudlift_version", "services", "notifications_arn"]
         }
         try:
+            for _, service_configuration in configuration.get('services').items():
+                for sidecar in service_configuration.get('sidecars', []):
+                    if sidecar.get('name') == FLUENTBIT_FIRELENS_SIDECAR_CONTAINER_NAME and sidecar.get('log_driver') == 'awsfirelens':
+                        raise UnrecoverableException("Logging set to awsfirelens, but fluentbit firelens sidecar container is missing or has incorrect log driver.")
```
Comment (shoan) [inline cloudlift/config/service_configuration.py:344]: This exception needs to be worded better

File: cloudlift/config/service_configuration.py [hunk 9]
```diff
@@ -279,7 +364,103 @@ def _default_service_configuration(self):
                     },
                     u'memory_reservation': 250,
                     u'command': None,
-                    u'spot_deployment': False
+                    u'spot_deployment': False,
+                    u'logging': 'awsfirelens'
```
Comment (shoan) [inline cloudlift/config/service_configuration.py:368]: We should default to AWS logs

File: cloudlift/deployment/service_template_generator.py [hunk 10]
```diff
@@ -254,7 +258,7 @@ def _add_service(self, service_name, config):
                 )
             ]
         if 'logging' not in config or 'logging' in config and config['logging'] is not None:
-            container_definition_arguments['LogConfiguration'] = self._gen_log_config(service_name, "awslogs" if 'logging' not in config else config['logging'])
+            container_definition_arguments['LogConfiguration'] = self._gen_log_config(service_name, "awsfirelens" if 'logging' not in config else config['logging'])
```
Comment (shoan) [inline cloudlift/deployment/service_template_generator.py:261]: Default needs to be aws logs

File: cloudlift/deployment/service_template_generator.py [hunk 11]
```diff
@@ -280,7 +284,20 @@ def _add_service(self, service_name, config):
                         Principal=Principal("Service", ["ecs-tasks.amazonaws.com"])
                     )
                 ]
-            )
+            ),
+            Policies=[
+                Policy(
+                    PolicyName=service_name + "-Firelens-" + hashlib.sha1(service_name.encode()).hexdigest()[0:5],
```
Comment (shoan) [inline cloudlift/deployment/service_template_generator.py:290]: Review the need for hashing

File: cloudlift/deployment/service_template_generator.py [hunk 12]
```diff
@@ -280,7 +284,20 @@ def _add_service(self, service_name, config):
...
+                    PolicyDocument=PolicyDocument(
+                        Statement=[
+                            Statement(
+                                Effect=Allow,
+                                Action=[PutRecordBatch, AssumeRole],
+                                Resource=["*"]
```
Comment (shoan) [inline cloudlift/deployment/service_template_generator.py:296]: Review to have more limited scope

File: cloudlift/deployment/service_template_generator.py [hunk 13]
```diff
@@ -549,10 +566,13 @@ def _gen_log_config(self, service_name, config):
...
                 }
             )
-        elif config == 'null':
+        elif config == 'awsfirelens':
             return LogConfiguration(
-                LogDriver="none"
+                LogDriver="awsfirelens"
             )
+        return LogConfiguration(
+            LogDriver="none"
```
Comment (shoan) [inline cloudlift/deployment/service_template_generator.py:574]: This should be the cluster default.

**Review by shoan — CHANGES_REQUESTED** [review]

File: cloudlift/deployment/service_template_generator.py (hunk 2 above)
Comment (rokibulislaam) [reply cloudlift/deployment/service_template_generator.py]: done

File: cloudlift/deployment/service_template_generator.py (hunk 4 above)
Comment (rokibulislaam) [reply cloudlift/deployment/service_template_generator.py:999]: removed `mount_points`

File: cloudlift/deployment/service_template_generator.py (hunk 5 above)
Comment (rokibulislaam) [reply cloudlift/deployment/service_template_generator.py:1018]: default has been set to `awslogs` for now

File: cloudlift/config/service_configuration.py (hunk 8 above)
Comment (rokibulislaam) [reply cloudlift/config/service_configuration.py:344]: reworded

File: cloudlift/deployment/service_template_generator.py (hunk 10 above)
Comment (rokibulislaam) [reply cloudlift/deployment/service_template_generator.py:261]: default logging has been set to `awslogs` for now

File: cloudlift/deployment/service_template_generator.py (hunk 11 above)
Comment (rokibulislaam) [reply cloudlift/deployment/service_template_generator.py:290]: Removed hashing, turned out to be redundant.

File: cloudlift/deployment/service_template_generator.py (hunk 12 above)
Comment (rokibulislaam) [reply cloudlift/deployment/service_template_generator.py:296]: Action `PutRecordBatch` has been removed. Changes have been made, from now on, the statement would look like this:

```py
Statement(
      Effect=Allow,
      Action=[AssumeRole],
      Resource=[assume_role_resource or "*"],
  )
# where assume_role_resource is the default role assume role ARN that has been on the environment configuration
```

File: cloudlift/deployment/service_template_generator.py (hunk 13 above)
Comment (rokibulislaam) [reply cloudlift/deployment/service_template_generator.py:574]: done

**Review by krishnanand275 — APPROVED** [review]
👍

**Review by shoan — APPROVED** [review]

**Pull request merged**